sklearn/cluster/hierarchical.py

```diff
@@ -743,14 +775,30 @@ def fit(self, X, y=None):
         if self.linkage != 'ward':
             kwargs['linkage'] = self.linkage
             kwargs['affinity'] = self.affinity
-        self.children_, self.n_components_, self.n_leaves_, parents = \
-            memory.cache(tree_builder)(X, connectivity,
-                                       n_clusters=n_clusters,
-                                       **kwargs)
+        distance_threshold = self.distance_threshold
+        # if distance_threshold is set then distances is returned
+        if distance_threshold is not None:
+            ch, n_comps, n_lvs, parents, distances = \
+                memory.cache(tree_builder)(X, connectivity,
+                                           n_clusters=n_clusters,
+                                           return_distance=True,
+                                           **kwargs)
+        else:
+            ch, n_comps, n_lvs, parents = \
+                memory.cache(tree_builder)(X, connectivity,
+                                           n_clusters=n_clusters,
+                                           **kwargs)
+        self.children_, self.n_components_, self.n_leaves_ = ch, n_comps, n_lvs
+
         # Cut the tree
         if compute_full_tree:
-            self.labels_ = _hc_cut(self.n_clusters, self.children_,
-                                   self.n_leaves_)
+            if distance_threshold is not None:
+                self.labels_ = _hc_cut(self.n_clusters, self.children_,
```

Comment:
why don't you just calculate n_clusters from distances here and avoid modifying `_hc_cut` (adding two parameters for the sake of 1 line of logic)?